### ISSUE TITLE
[sklearn#34596] Explicit documentation / contract for free-threading

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#312]
**Mirrored from [scikit-learn#34596](https://github.com/scikit-learn/scikit-learn/issues/34596)**

Original title: Explicit documentation / contract for free-threading
Original labels: Documentation, Needs Investigation, free-threading
Created: 2026-07-29T12:52:14Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the issue linked to the documentation

Scikit-learn offers builds for free-threaded Python, but it is not clear as a user what is supposed to be safe to call from multiple threads.

For example, it feels quite obvious that one should not call `.fit()` on the same object from multiple threads even if not documented, but does the interface guarantee that other methods are safe to call on the same object from multiple threads? For instance, KNN depending on the arguments might do most of the work in other methods, and I'm not sure if it doesn't modify any internal attributes starting with underscores.

Further, since lots of other Python libraries offer scikit-learn-compatible objects, it would be quite useful to have official documentation from scikit-learn about what needs or doesn't need to be thread-safe in order to develop scikit-learn-compatible classes.

### Suggest a potential alternative/fix

Add a documentation page for free threading covering all these kinds of details, plus additional documentation on a per-estimator basis if needed for exceptions. Mention these thread safety requirements also in the developer guide.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/111 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
